koku/koku/middleware.py

```diff
@@ -161,38 +184,87 @@ def process_request(self, request):
             DB_CONNECTION_ERRORS_COUNTER.inc()
             return HttpResponseFailedDependency({"source": "Database", "exception": err})
 
-    def get_tenant(self, model, hostname, request):
-        """Override the tenant selection logic."""
-        schema_name = "public"
+    def _check_user_has_access(self, request):
+        """Check if the user has access to Cost Management.
+
+        Args:
+            request (HttpRequest): The incoming request object.
+
+        Raises:
+            PermissionDenied: If the user does not have permissions for Cost Management.
+
+        """
+
+        username = request.user.username
+        if not request.user.admin and request.user.access is None:
+            LOG.warning("User %s does not have permissions for Cost Management.", username)
+            # For /user-access we do not want to raise the exception since the API will
+            # return a false boolean response that the platfrom frontend code is expecting.
+            if request.path != reverse("user-access"):
+                raise PermissionDenied()
+
+    def _get_or_create_tenant(self, request):
+        """Get or create tenant based on the user's schema.
+
+        Args:
+            request(HttpRquest): The incoming request object.
+
+        Returns:
+            Tenant: The tenant object.
+
+        """
+
+        if tenant := self._get_tenant_from_tenant_cache(request):
+            return tenant
+
+        tenant_username = request.user.username
+        schema_name = request.user.customer.schema_name if not is_no_auth(request) else "public"
+        if tenant := self._get_tenant_from_db(tenant_username, schema_name):
+            return tenant
+
+        return self._create_tenant()
+
+    def _get_tenant_from_tenant_cache(self, request):
+        """Get tenant from tenant cache."""
+
         tenant_username = request.user.username
         tenant = KokuTenantMiddleware.tenant_cache.get(tenant_username)
-        if not tenant:
-            if not is_no_auth(request):
-                schema_name = request.user.customer.schema_name
+        return tenant
 
-            tenant = model.objects.filter(schema_name=schema_name).first()
-            if tenant and schema_name != "public":
+    def _get_tenant_from_db(self, tenant_username, schema_name):
+        """Get tenant with the schema from the database."""
+
+        try:
+            tenant = Tenant.objects.get(schema_name=schema_name)
+            if schema_name != "public":
                 with KokuTenantMiddleware.tenant_lock:
                     KokuTenantMiddleware.tenant_cache[tenant_username] = tenant
                     LOG.debug(f"Tenant added to cache: {tenant_username}")
-            elif not tenant:
-                tenant, __ = model.objects.get_or_create(schema_name="public")
+            return tenant
+        except Tenant.DoesNotExist:
+            LOG.info(f"Tenant does not exist: {tenant_username}")
+            return
```

Comment:
I also think this method should `raise` rather than returning `None`. That is [mode idiomatic](https://docs.quantifiedcode.com/python-anti-patterns/readability/asking_for_permission_instead_of_forgiveness_when_working_with_files.html).

Reply:
Thanks for this suggestion. 
I think this also would help with troubleshooting if this exception were to be raised. This is now updated.